flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,454 @@
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Tests for the partition-lifecycle logic in the {@link TaskExecutor}.
+ */
+public class TaskExecutorPartitionLifecycleTest extends TestLogger {
+
+	private static final Time timeout = Time.seconds(10L);
+
+	private static final TestingHighAvailabilityServices haServices = new TestingHighAvailabilityServices();
+	private static final SettableLeaderRetrievalService jobManagerLeaderRetriever = new SettableLeaderRetrievalService();
+	private static final JobID jobId = new JobID();
+
+	static {
+		haServices.setResourceManagerLeaderRetriever(new SettableLeaderRetrievalService());
+		haServices.setJobMasterLeaderRetriever(jobId, jobManagerLeaderRetriever);
+	}
+
+	@Rule
+	public final TemporaryFolder tmp = new TemporaryFolder();
+
+	private TestingRpcService rpc;
+
+	@Before
+	public void setup() {
+		rpc = new TestingRpcService();
+	}
+
+	@After
+	public void shutdown() throws ExecutionException, InterruptedException {
+		if (rpc != null) {
+			rpc.stopService().get();
+		}
+	}
+
+	@Test
+	public void testConnectionTerminationAfterExternalRelease() throws IOException, InterruptedException, ExecutionException, TimeoutException {
+		final JobMasterGateway jobMasterGateway = new TestingJobMasterGatewayBuilder().build();
+
+		final JobManagerConnection jobManagerConnection = TaskSubmissionTestEnvironment.createJobManagerConnection(
+			jobId, jobMasterGateway, rpc, new NoOpTaskManagerActions(), timeout);
+
+		final JobManagerTable jobManagerTable = new JobManagerTable();
+		jobManagerTable.put(jobId, jobManagerConnection);
+
+		final AtomicBoolean hasPartitionsOccupyingLocalResources = new AtomicBoolean(true);
+		final TestJobAwareShuffleEnvironment jobAwareShuffleEnvironment = new TestJobAwareShuffleEnvironment(jobId -> hasPartitionsOccupyingLocalResources.get());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setJobManagerTable(jobManagerTable)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.setTaskSlotTable(createTaskSlotTable())
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices);
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			assertTrue(jobManagerTable.contains(jobId));
+
+			taskManager.releasePartitions(jobId, Collections.singletonList(new ResultPartitionID()));
+			// connection should be kept alive since the environment still says we have local resources
+			assertTrue(jobManagerTable.contains(jobId));
```

Comment:
This check has concurrency issues becuase `jobManagerTable` is modified both from the `TaskExecutor` and checked here. I think it would be better to check the closing on the JM connection via the `TestingJobMasterGateway` by setting a `disconnectTaskManagerConsumer`.